pkg/controller/job/job_controller_test.go

```diff
@@ -1916,6 +1920,7 @@ func TestPastDeadlineJobFinished(t *testing.T) {
 				if len(j.Status.Conditions) == 1 && j.Status.Conditions[0].Reason == "DeadlineExceeded" {
 					return true, nil
 				}
+				manager.clock.Sleep(time.Second)
```

Comment:
is there a way to access the prequeue? https://github.com/kubernetes/kubernetes/issues/110697#issuecomment-1378760495

Reply:
Does not seem to be exported, maybe by reflection.

Reply:
Another idea: add `AddAt`, in addition to `AddAfter` to the `DelayingInterface`.

It's weird that we have to calculate the remaining time just to add it back to the current time here again:

https://github.com/kubernetes/client-go/blob/674e53a7cb8f1ea160ac4485ffc7582e1a790215/util/workqueue/delaying_queue.go#L179

Reply:
AddAt sounds like a nice function, but would it help to fix the test, I guess it calls for another pr.

Reply:
yes, it would, because we precisely set the time, instead of it being based on the current time in the clock in the workqueue.

Reply:
I want to change the job_controller to call `AddAt`.

> I think the issue is that AddAfter can be called before or after we move the time in job_controller_test.

Exactly. The test fails when `AddAfter` is called after we move the time. This could still happen with `AddAt`, but since the time already passed, `AddAt` would just put it in the queue directly.

Reply:
I see now, just pushed that change. PTAL.

Reply:
Ah, there are still flakes after this change, don't understand it yet, here is the example log with debugs:

```
3987463-=== RUN   TestPastDeadlineJobFinished/New_job_created_without_start_time_being_set
3987546-I0112 14:49:36.776560  772603 job_controller.go:199] Starting job controller
3987623-I0112 14:49:36.776578  772603 shared_informer.go:314] Waiting for caches to sync for job
3987712-I0112 14:49:36.776599  772603 shared_informer.go:321] Caches are synced for job
3987792-I0112 14:49:36.776677  772603 job_controller.go:516] enqueueing job default/job1
3987873-I0112 14:49:36.776753  772603 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 14:49:36 +0100 CET"
3987993-I0112 14:49:36.776810  772603 job_controller.go:794] "MYDEBUG" past=false timeoutAt="2023-01-12 14:49:37 +0100 CET"
3988109-I0112 14:49:36.776980  772603 job_controller.go:542] "MYDEBUG before done" queue.Len()=0
3988198-I0112 14:49:36.776997  772603 job_controller.go:544] "MYDEBUG after done" queue.Len()=0
3988286-I0112 14:49:36.777010  772603 job_controller.go:516] enqueueing job default/job1
3988367-I0112 14:49:36.777017  772603 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 14:49:36 +0100 CET"
3988487-I0112 14:49:36.777033  772603 job_controller.go:794] "MYDEBUG" past=false timeoutAt="2023-01-12 14:49:37 +0100 CET"
3988603-    job_controller_test.go:1910: Time before sleep: 2023-01-12 14:49:36 +0100 CET, 0
3988688-    job_controller_test.go:1912: Time after sleep: 2023-01-12 14:49:37 +0100 CET, 0
3988949-    job_controller_test.go:1925: Job failed to enforce activeDeadlineSeconds configuration. Expected condition with Reason 'DeadlineExceeded' was not found in {[] 2023-01-12 14:49:36 +0100 CET <nil> 1 0 0  &UncountedTerminatedPods{Succeeded:[],Failed:[],} 0xc000ace9ac}
3989219:--- FAIL: TestPastDeadlineJobFinished (3.10s)

```

Also tested, that after this change there are no flakes again if I move the clock by 1s.

I suspect that even if we add to the queue, but there is no time passage then the item is never picked up, but don't really understand it yet, so might be wrong.

Reply:
This is an example run with added sleep(200*time.Milliseconds), that would fail if not for the sleep. 

```
1579771-I0112 15:23:07.384932  784463 job_controller.go:213] Shutting down job controller
1579853-=== RUN   TestPastDeadlineJobFinished/New_job_created_without_start_time_being_set
1579936-I0112 15:23:07.485768  784463 job_controller.go:199] Starting job controller
1580013-I0112 15:23:07.485809  784463 shared_informer.go:314] Waiting for caches to sync for job
1580102-I0112 15:23:07.485824  784463 shared_informer.go:321] Caches are synced for job
1580182-I0112 15:23:07.485886  784463 job_controller.go:516] enqueueing job default/job1
1580263-I0112 15:23:07.485950  784463 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 15:23:07 +0100 CET"
1580383-I0112 15:23:07.486003  784463 job_controller.go:794] "MYDEBUG" past=false timeoutAt="2023-01-12 15:23:08 +0100 CET"
1580499-I0112 15:23:07.486165  784463 job_controller.go:542] "MYDEBUG before done" queue.Len()=0
1580588-I0112 15:23:07.486183  784463 job_controller.go:544] "MYDEBUG after done" queue.Len()=0
1580676-I0112 15:23:07.486198  784463 job_controller.go:516] enqueueing job default/job1
1580757-I0112 15:23:07.486209  784463 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 15:23:07 +0100 CET"
1580877-I0112 15:23:07.486228  784463 job_controller.go:794] "MYDEBUG" past=false timeoutAt="2023-01-12 15:23:08 +0100 CET"
1580993-    job_controller_test.go:1910: MYDEBUG Time before sleep: 2023-01-12 15:23:07 +0100 CET, 0
1581086-    job_controller_test.go:1912: MYDEBUG Time after sleep: 2023-01-12 15:23:08 +0100 CET, 0
1581178-I0112 15:23:07.486416  784463 job_controller.go:542] "MYDEBUG before done" queue.Len()=0
1581267-I0112 15:23:07.486445  784463 job_controller.go:544] "MYDEBUG after done" queue.Len()=0
1581355:    job_controller_test.go:1921: MYDEBUG Looping: 2023-01-12 15:23:08 +0100 CET, 0
1581438:    job_controller_test.go:1921: MYDEBUG Looping: 2023-01-12 15:23:08.2 +0100 CET, 0
1581523:    job_controller_test.go:1921: MYDEBUG Looping: 2023-01-12 15:23:08.4 +0100 CET, 0
1581608:    job_controller_test.go:1921: MYDEBUG Looping: 2023-01-12 15:23:08.6 +0100 CET, 0
1581693:    job_controller_test.go:1921: MYDEBUG Looping: 2023-01-12 15:23:08.8 +0100 CET, 0
1581778-I0112 15:23:08.487010  784463 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 15:23:09 +0100 CET"
1581898-I0112 15:23:08.487069  784463 job_controller.go:796] "MYDEBUG" past=true timeoutAt="nil"
1581987-I0112 15:23:08.487170  784463 job_controller.go:542] "MYDEBUG before done" queue.Len()=0
1582076-I0112 15:23:08.487178  784463 job_controller.go:544] "MYDEBUG after done" queue.Len()=0
1582164-I0112 15:23:08.487258  784463 event.go:294] "Event occurred" object="default/job1" fieldPath="" kind="Job" apiVersion="" type="Warning" reason="DeadlineExceeded" message="Job was active longer than specified deadline"
1582382-I0112 15:23:08.487266  784463 job_controller.go:516] enqueueing job default/job1
1582463-I0112 15:23:08.487297  784463 job_controller.go:516] enqueueing job default/job1
1582544-I0112 15:23:08.487323  784463 job_controller.go:678] "MYDEBUG syncJob starts" startTime="2023-01-12 15:23:09 +0100 CET"
1582664-I0112 15:23:08.487373  784463 job_controller.go:542] "MYDEBUG before done" queue.Len()=0
1582753-I0112 15:23:08.487386  784463 job_controller.go:544] "MYDEBUG after done" queue.Len()=0
```
So it is clearly AddAt invoked with `"2023-01-12 15:23:08 +0100 CET"`, but it is invoked for some reason only once the time passes to 2023-01-12 15:23:09 +0100 CET

Reply:
But if AddAt is invoked with past the time, the item should be directly added to the workqueue.
Maybe there's something wrong with the `AddAt` implementation.

Reply:
Ok, I think I understand now why AddAt does not solve the flakes (but the scenario is more subtle than for AddAfter). This is the scenario:
- Test: creates a job with ActiveDeadlineSeconds=1 and clock at 0
- Job: sets the job.Status.StartTime=0
- Job: enters pastActiveDeadline and calls AddAt(1)
- DelayingQueue: add the item to waitingForAddCh as it is still at 0 time: https://github.com/kubernetes/kubernetes/blob/01d44159157529f8354e475ec796712298ca6608/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L204
- DelayingQueue: peeks up the entry and inserts into waitingForQueue https://github.com/kubernetes/kubernetes/blob/a66aad2d80dacc70025f95a8f97d2549ebd3208c/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L245
- DelayingQueue: peeks up the entry from waitingForQueue and decides to create the timer, determines the timer for 1s: https://github.com/kubernetes/kubernetes/blob/a66aad2d80dacc70025f95a8f97d2549ebd3208c/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L229
- Test: move the clock to 1s
- DelayingQueue: create the timer for 1s, but the timer will write to the channel at 2: 
https://github.com/kubernetes/kubernetes/blob/a66aad2d80dacc70025f95a8f97d2549ebd3208c/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L240
- Issue: we never move the clock to 2, but it also explains why moving the clock to 2 in the test helps

The inner workings of the delaying_queue are sensitive to the time moving forward, because they create a new inner timer based on the time difference (exposing to similar issues as AddAfter). 

It isn't clear to me how to "fix" that, and if the effort is worth the gain, given we have a one-liner solution (or a workaround) of adding `manager.clock.Sleep(200ms)`. The only downside of this I can think of is that occasionally the test would take 1s longer, but it is probably less than 1% of runs.

Reply:
A potential solution would be to have a `clock.NewTimerAt`, but that deviates too much from the standard `time` package.


What we can do is to remove the `clock.Sleep(1s)` and only leave the 200ms increments. But we should probably also throw an error if the job fails before the 1s is completed.

And we can get rid of `AddAt`, if we can't guarantee its behavior.